Bonsai.Shaders/BindTexture.cs

```diff
@@ -76,43 +92,25 @@ IObservable<TSource> Process<TSource>(IObservable<TSource> source, Action<TSourc
                                 : null;
                         }
 
+                        texture ??= input as Texture;
```

Comment:
> I see one potentially problematic edge-case: if you didn't specify a texture name, and you pass a sequence of valid textures, followed by a null texture, the `texture` state variable will bind to the latest non-null texture.

`BindTexture`: The most nuanced operator in all of Bonsai.

> (will this be optimized by the compiler when generic type will always fail the test? or the other way around, are we paying the cost of boxing to object by using as with generic types in cases where the actual type is a value type?)

For reference types it's implementation-defined whether the runtime will specialize a generic method for each `T` or if it will perform generic sharing. Not sure if it would in this case, my gut says it wouldn't since historically the JIT has been pretty conservative about doing so. (I'd check but Disasmo only works with modern .NET and it seems overkill to go mucking with WinDbg for this.) Generic sharing should always be avoided when inlining happens, but inlining almost certainly won't happen in this case since the method is fat and tail calls make more sense.

For value types it will always specialize each T out of necessity. (The .NET VM makes strong guarantees about how value types are treated that inevitably require it.)

It's mostly a moot point though, `input` will only be inspected when `texture` is `null`, it's only `null` if the node isn't configured. I don't think it's valuable to sweat the small stuff for a `BindTexture` node which isn't configured and doesn't take `Texture` elements. (The only reasonable case I can think of this happening is if `TextureName` is externalized and someone sets it to `null`, but this doesn't actually do anything useful since the unbinding is skipped in that situation.)

> on the other hand, if we know the type is a texture we could just do a direct type-cast rather than `as`.

This doesn't ever have any meaningful performance difference as far as I'm aware, they're very similar operations because both have to handle the cast being wrong. (And that check would only be elided if the generic was specialized, which would've also removed the similar handling for `as` anyway.)

> getter delegate might be faster

The getter delegate has the overhead of allocating the closure on the heap. Invoking it isn't gonna be super friendly towards the data/instruction cache or branch predictor. I'd be shocked if it was consistently faster over casting :)

-----------------

Anyway your code as suggested wasn't going to fix the missing handling of `null`-to-unbind.

I pushed a new iteration (hopefully the final one lol) which properly handles `null` in the texture-from-input case and avoids incorrectly caching the input texture.